src/components/products/product-subscribe/react/ProductSubscribeRuntime.tsx

```diff
@@ -0,0 +1,182 @@
+                    return;
+                }
+
+                // TODO better error handling & logging
+                if (error.code === "ResourceNotFound") {
+                    return;
+                }
+
+                throw new Error(`Unable to load products. Error: ${error.message}`);
+            })
+            .finally(() => setWorking(false));
+    }, [usersService, tenantService, productService, productName]);
+
+    const hasToS = !!product?.terms;
+    const subscribe: TSubscribe = useCallback(async (subscriptionName: string, consented?: boolean) => {
+        const canSubscribe = (isDelegationEnabled || subscriptionName.length > 0) && (!hasToS || consented);
+        if (!productName || !userId || !canSubscribe) return;
+
+        return handleSubscribing(backendService, productService, productName, userId, subscriptionName, isDelegationEnabled)
+            .then(() => usersService.navigateToProfile())
+            .catch((error) => {
+                if (error.code === "Unauthorized") {
+                    usersService.navigateToSignin();
+                    return;
+                }
+                // TODO better error handling & logging
+                throw new Error(`Unable to subscribe to a product. Error: ${error.message}`);
+            })
+    }, [backendService, productService, productName, userId, isDelegationEnabled, hasToS]);
+
+    if (working) return <Spinner size="extra-tiny" label={"Loading data"} labelPosition={"after"} />;
+    if (isLimitReached) return <Body1Strong style={{ display: "block", padding: "1rem 0" }}>You've reached maximum number of subscriptions.</Body1Strong>;
+
+    return <ProductSubscribeForm subscribe={subscribe} tos={product?.terms} showTermsByDefault={showTermsByDefault} />;
+};
+
+export class ProductSubscribeRuntime extends React.Component<
+    ProductSubscribeRuntimeProps,
+    { productName?: string | null }
+> {
+    @Resolve("backendService")
+    public backendService: BackendService;
+
+    @Resolve("usersService")
+    public usersService: UsersService;
+
+    @Resolve("productService")
+    public productService: ProductService;
+
+    @Resolve("tenantService")
+    public tenantService: TenantService;
+
+    @Resolve("routeHelper")
+    public routeHelper: RouteHelper;
+
+    @Resolve("router")
+    public router: Router;
+
+    constructor(props: ProductSubscribeRuntimeProps) {
+        super(props);
+
+        this.state = { productName: undefined };
+    }
+
+    componentDidMount() {
+        this.setProductName();
+        this.router.addRouteChangeListener(this.setProductName.bind(this));
+    }
+
+    componentWillUnmount() {
+        this.router.removeRouteChangeListener(this.setProductName.bind(this));
+    }
+
+    setProductName() {
+        this.setState({ productName: this.routeHelper.getProductName() });
+    }
+
+    render() {
+        if (!this.state.productName) return <>Please select a product</>;
```

Comment:
Body1 tag?

Reply:
as it's outside of FluentProvider it wouldn't work and this should never happen for C2

Reply:
Oh, my bad.